SNI-based cert selection during TLS handshake (#22036)
envoyproxy/envoy: docs/root/intro/arch_overview/security/ssl.rst

---------------------

:ref:`DownstreamTlsContexts <envoy_v3_api_msg_extensions.transport_sockets.tls.v3.DownstreamTlsContext>` support multiple TLS
certificates. These may be a mix of RSA and P-256 ECDSA certificates for multiple services. The following rules apply:

Certificate config/loading rules:

* Only one certificate of a particular type (RSA or ECDSA) may be specified.
* Non-P-256 server ECDSA certificates are rejected.
* Static and SDS certificates may not be mixed in a given :ref:`DownstreamTlsContext
  <envoy_v3_api_msg_extensions.transport_sockets.tls.v3.DownstreamTlsContext>`.

SNI matching rules:

* If the client support SNI, a certificate with proper DNS SANs or Subject Common Name should be selected

> * If no certificate is matched to SNI, the connection is refused.

Author: Got it.

> * If no certificate is matched to SNI, the connection is refused.

Author: @ggreenway Since we introduce SNI-based selection, we can't restrict there are 1 or 2 certs in total. We should restrict 1 or 2 certs(RSA, ESDSA) for each Server Name Pattern (DNS SAN entry or Subject Common Name). See code and doc for details.

Certificate config/loading rules:

* DNS SANs or Subject Common Name is extracted as server name pattern to match SNI during handshake.
* FQDN like "test.example.com" and wildcard like "\*.example.com" are valid at the same time, which will be loaded
  as two different server name pattern.
* Only one certificate of a particular type (RSA or ECDSA) may be specified for each server name pattern.
* Non-P-256 server ECDSA certificates are rejected.
* Static and SDS certificates may not be mixed in a given :ref:`DownstreamTlsContext
  <envoy_v3_api_msg_extensions.transport_sockets.tls.v3.DownstreamTlsContext>`.

SNI matching rules:

* If the client support SNI, a certificate with proper DNS SANs or Subject Common Name should be selected
* It tries to match on exact server name first, then match on wildcard server name. e.g. If SNI is
  "test.example.com", a group of certificates with "test.example.com" will become candidates if it is present,
  otherwise it looks for ".example.com".
* If no certificate is matched to SNI or the client does not support SNI, subsequent particular type (RSA or ECDSA)
  matching will be executed with all certificates as candidates.

> * Otherwise, particular type (RSA or ECDSA) matching will be executed with SNI-matched certificates as candidates.

Reviewer: But with the current ordering, the docs are intermixing the SNI-match and no-match cases. If you put this point before the previous, all the SNI-match docs come before the non-match. I think that makes it easier to understand.

> * Otherwise, particular type (RSA or ECDSA) matching will be executed with SNI-matched certificates as candidates.

Author: Got your point, will update

Public Key Type matching(ECDSA or RSA) rules:

* If the client supports P-256 ECDSA, a P-256 ECDSA certificate is selected if it is present and the OCSP check is passed.
* If the client only supports RSA, a RSA certificate is selected if it is present.
* If no exact match, fallback to the first certificate in the candidates.
* The certificate that it fallbacks to might result in a failed handshake. For instance, a client only supports

>   RSA certificates and the certificate only support ECDSA, or a client only supports ECDSA certificate and the

Reviewer: I'm honestly not sure. But if it's possible, it's rare enough that I wouldn't call out that example in the docs. But keep the other example of a client that supports only RSA.

>   RSA certificates and the certificate only support ECDSA, or a client only supports ECDSA certificate and the

Author: Got it, thanks.

---------------------

:ref:`DownstreamTlsContexts <envoy_v3_api_msg_extensions.transport_sockets.tls.v3.DownstreamTlsContext>` support multiple TLS
certificates. These may be a mix of RSA and P-256 ECDSA certificates for multiple SANs.

Certificate config/loading rules:

* DNS SANs or Subject Common Name is extracted as server name pattern to match SNI during handshake.
* FQDN like "test.example.com" and wildcard like "\*.example.com" are valid at the same time, which will be loaded
  as two different server name pattern.
* Only one certificate of a particular type (RSA or ECDSA) may be specified for each server name pattern.
* Non-P-256 server ECDSA certificates are rejected.
* Static and SDS certificates may not be mixed in a given :ref:`DownstreamTlsContext
  <envoy_v3_api_msg_extensions.transport_sockets.tls.v3.DownstreamTlsContext>`.

Certificate selection rules, including SNI matching, Public Key Type matching and OCSP:

* If the client supports SNI, e.g. SNI is "test.example.com", it looks for a cert that exactly matches to the SNI.
* If no cert is selected from certs that exactly matches SNI, It matches on wildcard server name. e.g. if SNI is

>   "test.example.com", a certificate with "test.example.com" will be preferred over "\*.example.com".

Author: done

* If no cert is selected from certs that matches wildcard name, it looks for a cert from whole cert list.
* It requires OCSP and key type checking in above phases, the candidate cert can be selected only when it adhere to the OCSP policy.
  It tries to find a cert that can both match OCSP policy and key type, if such cert is not present, the first cert that passing OCSP
  will be selected even if it does not match to key type.
* It fallbacks to the first cert in the whole list if there is no cert is selected after above steps.

> * Currently only two kinds of key type are supported, RSA or ECDSA, If the client supports P-256 ECDSA, the P-256 ECDSA certificate

Reviewer: ```suggestion
* Currently only two kinds of key type are supported, RSA or ECDSA. If the client supports P-256 ECDSA, the P-256 ECDSA certificate
```

>   is preferred over RSA. The certificate that it fallbacks to might result in a failed handshake. For instance, a client only supports

Author: done

---------------------

:ref:`DownstreamTlsContexts <envoy_v3_api_msg_extensions.transport_sockets.tls.v3.DownstreamTlsContext>` support multiple TLS
certificates. These may be a mix of RSA and P-256 ECDSA certificates for multiple SANs.

Certificate config/loading rules:

* DNS SANs or Subject Common Name is extracted as server name pattern to match SNI during handshake.
* FQDN like "test.example.com" and wildcard like "\*.example.com" are valid at the same time, which will be loaded
  as two different server name pattern.
* Only one certificate of a particular type (RSA or ECDSA) may be specified for each server name pattern.
* Non-P-256 server ECDSA certificates are rejected.

> * Static and SDS certificates may not be mixed in a given :ref:`DownstreamTlsContext

Reviewer: Does multiple certs via SDS work the same as multiple static certs?

Author: Yes. And we are working on cert provider, it will work as well.

Reviewer: Excellent.

  <envoy_v3_api_msg_extensions.transport_sockets.tls.v3.DownstreamTlsContext>`.

Certificate selection rules, including SNI matching, Public Key Type matching and OCSP:

> * If the client supports SNI, a certificate with proper DNS SANs or Subject Common Name should be selected

Reviewer: ```suggestion
* If the client supports SNI, a certificate with proper DNS SANs or Subject Common Name should be selected.
```

* It matches on exact server name first, then matches on wildcard server name if an exact name match isn't found, e.g. if SNI is

>   "test.example.com", a certificate with "test.example.com" will be preferred over for "\*.example.com".

Reviewer: ```suggestion
  "test.example.com", a certificate with "test.example.com" will be preferred over "\*.example.com".
```

> * If no certificate is matched to SNI or the client does not support SNI, subsequent particular type (RSA or ECDSA)
>   matching will be executed with all certificates as candidates.

Reviewer: ```suggestion
* If no certificate is matched to SNI or the client does not support SNI, key type (RSA or ECDSA)
  matching will be executed with all certificates as candidates.
```

> * Otherwise, particular type (RSA or ECDSA) matching will be executed with SNI-matched certificates as candidates.

Reviewer: ```suggestion
* Otherwise, key type (RSA or ECDSA) matching will be executed with SNI-matched certificates as candidates.
```

* Key type matching is based on SNI matching results
* If the client supports P-256 ECDSA, the first P-256 ECDSA certificate is selected if it is present and the OCSP check is passed.
* If the client only supports RSA, the first RSA certificate is selected if it is present.
* If no exact match, fallback to the first certificate in the candidates.
* The certificate that it fallbacks to might result in a failed handshake. For instance, a client only supports
  RSA certificates and the certificate only support ECDSA, or a client only supports ECDSA certificate and the
  certificate only support RSA.

> * After key type matching, one certificate is selected, and the selected certificate must adhere to the OCSP policy.

Reviewer: This is a change from the previous behavior. I think you need to retain previous behavior. The previous docs specify that a cert is only selected "if it is in compliance with the OCSP policy." 

To be consistent, we should probably check this while doing SNI matching, so that if test.example.com cert is out of OCSP policy but *.example.com is in policy, *.example.com is selected.

Author: we do not break previous docs, currently if ocsp fails, we return ssl_select_cert_error.

I think fallback to wildcard SNI matching due to OCSP fail brings some complexity, what if pkey matching fails? Do we need to look up wildcard SNI matching again. Which cert should we fallback to, the first exact SNI-matched certs, the first wildcard SNI-matched certs, or first cert without any match?

Author: I wrote alternative code to support wildcard SNI matching if ocsp/pkey matching fails with exact SNI, and we will always fallback to the first cert in the tls context list.

```cpp
  auto selected = [&](const TlsContext& ctx) -> bool {
    if (client_ecdsa_capable != ctx.is_ecdsa_) {
      return false;
    }

    auto action = ocspStapleAction(ctx, client_ocsp_capable);
    if (action == OcspStapleAction::Fail) {
      return false;
    }

    selected_ctx = &ctx;
    ocsp_staple_action = action;
    return true;
  };

  auto select_from_map = [&]() -> void {
    if (server_name_it == server_names_map_.end()) {
      return;
    }
    const auto& pkey_types_map = server_name_it->second;
    for (const auto& entry : pkey_types_map) {
      if (selected(entry.second.get())) {
        break;
      }
    }
  };

  // Do SNI matching and pkey type matching if SNI exists.
  if (!sni.empty()) {
    // Match on exact server name, i.e. "www.example.com" for "www.example.com".
    server_name_it = server_names_map_.find(sni);
    select_from_map();

    if (selected_ctx == nullptr) {
      // Match on wildcard domain, i.e. ".example.com" for "www.example.com".
      // https://datatracker.ietf.org/doc/html/rfc6125#section-6.4
      size_t pos = sni.find('.', 1);
      if (pos < sni.size() - 1 && pos != std::string::npos) {
        absl::string_view wildcard = sni.substr(pos);
        server_name_it = server_names_map_.find(static_cast<std::string>(wildcard));
      }
      select_from_map();
    }
  }
  // Do pkey type matching if SNI does Not exist or no ctx matched for SNI.
  if (selected_ctx == nullptr) {
    // Fallback on first certificate.
    selected_ctx = &tls_contexts_[0];
    ocsp_staple_action = ocspStapleAction(*selected_ctx, client_ocsp_capable);
    for (const auto& ctx : tls_contexts_) {
      if (selected(ctx)) {
        break;
      }
    }
  }
```

Reviewer: Yes, I think something like this is the correct way to do it. This prevents a partial match/selection on exact name from blocking a full match/selection on wildcard.

But to make that code snippet clearer, I think you should pass the iterator as a parameter to the lambda, and have the lambda only capture `this`.

Author: OK, will update the PR.

* It requires OCSP and key type checking in above phases, the candidate cert can be selected only when it adhere to the OCSP policy.
  It tries to find a cert that can both match OCSP policy and key type, if such cert is not present, the first cert that passing OCSP
  will be selected even if it does not match to key type.
* It fallbacks to the first cert in the whole list if there is no cert is selected after above steps.
* Currently only two kinds of key type are supported, RSA or ECDSA. If the client supports P-256 ECDSA, the P-256 ECDSA certificate
  is preferred over RSA. The certificate that it falls back to might result in a failed handshake. For instance, a client only supports
  RSA certificates and the certificate only support ECDSA.
* The final selected certificate must adhere to the OCSP policy. If no such certificate is found, the connection is refused.

.. note::
   With the support of SNI-based certificate selection, it allows configuring large number of certificates for multiple hostnames.
   But if there is no matching SNI or there is OCSP policy failure on the matching SNI, it will look for the cert from the whole cert

>    list(O(n) complexity), which could be a problem for a potential DoS attack.

Reviewer 2: I discussed this with @ggreenway and we think the following would be a better/safer default for both DoS safety in Envoy as well as more secure behavior for clients.

1) If no SNI is provided by the client, the behavior should be the same as today. (Adhere to OSCP, etc.)
2) If a client provides SNI, there must be either an exact or wildcard match. If there is no such match, the behavior would be as follows:
  a) Consult a new configuration option which specifies whether to abort the handshake or proceed with a legacy style fulls scan match.
  b) The default for the configuration would be false (abort and do not do a full scan)
  c) we should use a WKT for the boolean, and we should provide a runtime guard that can override the default in case people run into trouble with the default option.

Let me know if this makes sense and @ggreenway lmk if I missed anything.

Author: @mattklein123 thanks for your comments

> 1. If no SNI is provided by the client, the behavior should be the same as today. (Adhere to OSCP, etc.)
It can not be exact "the same ". Since we allow loading more than 2 certs with new change, If there is no SNI, this becomes O(n) lookup.  @ggreenway and me considered limit the times of lookup, e.g. we only allow it check the first 2 certs, it can avoid dos issue though it looks not elegant.

> 2. If a client provides SNI, there must be either an exact or wildcard match. If there is no such match, the behavior would be as follows:
>    a) Consult a new configuration option which specifies whether to abort the handshake or proceed with a legacy style fulls scan match.
>    b) The default for the configuration would be false (abort and do not do a full scan)
>    c) we should use a WKT for the boolean, and we should provide a runtime guard that can override the default in case people run into trouble with the default option.

a new configuration option makes sense. What does WKT mean in this context?

Reviewer 2: Sorry WKT means "well known type" so instead of bool you should use https://developers.google.com/protocol-buffers/docs/reference/google.protobuf#google.protobuf.BoolValue so you can differentiate between unset/false/true.

Author: Thanks.

Are you OK with full scan when no SNI is provided by the client? 

Regarding the the new option, I think it should reside in [extensions.transport_sockets.tls.v3.DownstreamTlsContext](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/transport_sockets/tls/v3/tls.proto#extensions-transport-sockets-tls-v3-downstreamtlscontext), do you agree?
```protobuf
message DownstreamTlsContext {
...
// default is false
google.protobuf.BoolValue full_scan_certs_on_sni_mismatch = 8;
}
```

Reviewer: The difference is that previously we said you should only have 2 certs, so full scan was O(2). 

The only realistic case I worry about is if a config only has EC certs, and has LOTS of them, and a client sends a ClientHello specifying only RSA is supported, and no SNI value. We could easily work around that by having a bool computed at config-load for `has_ec` and `has_rsa`, and check that first before even looking for a matching cert (regardless of SNI or not). We can quickly bail out when we know we don't have a compatible cert.

Reviewer 2: > The difference is that previously we said you should only have 2 certs, so full scan was O(2).

Author: > We said that but did we check that? If we actually did let people have more than 2 previously maybe it doesn't matter?

We did check that. Envoy would exit when it detected more than 2 certs, at most one ECDSA and at most one RSA was allowed since Envoy only supports these two types. In this PR, we allow at most 2 certs(one ECDSA and one RSA) for each server name pattern, so we might have 2 certs for www.example1.com, 2 for www.example2.com, 2 for *.example.com, and more. Though it is always O(n) lookup, but n <= 2 previously.

Reviewer: I'd rather use the existing field. The new behavior is close enough to the old behavior that I don't think we need a new field.

Reviewer 2: Alright sgtm. I guess let's just defer to trying to avoid any full scans by default.

Author: @ggreenway @mattklein123 please review the code update, it should meet all requirments.

certificates. These may be a mix of RSA and P-256 ECDSA certificates for multiple SANs.

Certificate config/loading rules:

* DNS SANs or Subject Common Name is extracted as server name pattern to match SNI during handshake. Subject Common Name is not used if DNS SANs are present in the certificate.
* FQDN like "test.example.com" and wildcard like "\*.example.com" are valid at the same time, which will be loaded
  as two different server name patterns.
* Only one certificate of a particular type (RSA or ECDSA) may be specified for each server name pattern.
* Non-P-256 server ECDSA certificates are rejected.
* Static and SDS certificates may not be mixed in a given :ref:`DownstreamTlsContext
  <envoy_v3_api_msg_extensions.transport_sockets.tls.v3.DownstreamTlsContext>`.

Certificate selection rules:

* If the client supports SNI, e.g. SNI is "test.example.com", it looks for a cert that exactly matches to the SNI.
  If the certificate adheres to the OCSP policy and matches to key type, it is selected for handshake.
  If the certificate adheres to the OCSP policy, but key type is RSA while client is ECDSA capable, it is marked as

>   as the candidate and continues searching until a cert is selected with perfect match or certs exhausted.

Reviewer 2: Is this part gated by the full scan option? If so can you clarify?

Author: No, this part is not gated by full scan option.

  Candidate will be selected for handshake if there is no perfect match.
* If the client supports SNI, but no cert is selected from certs that exactly matches to SNI, it matches on wildcard server name.
  e.g. if SNI is "test.example.com", a certificate with "test.example.com" will be preferred over "\*.example.com". And wildcard
  matching only works for 1 level of depth, so "\*.com" will not be a match for "test.example.com".
  Afterwards, it execuates OCSP and key type checking on each cert which is the same as what happens after exact SNI matching.
* If no cert is selected from certs that matches wildcard name, the candidate cert is selected for handshake if it is present.
  If there is no candidate, check :ref:`full_scan_certs_on_sni_mismatch <envoy_v3_api_field_extensions.transport_sockets.tls.v3.DownstreamTlsContext.full_scan_certs_on_sni_mismatch>`,
  go to full scan all certificates if it is enabled, otherwise pick the first certificate for handshake.

> * If the client does not provide SNI, go to full scan by default.

Reviewer 2: This is gated by the config option, correct? Clarify?

* Full scan execuates OCSP and key type checking on each cert which is the same as described above in exact SNI matching.
  It falls back to the first cert in the whole list if there is no cert selected.
* Currently only two kinds of key type are supported, RSA or ECDSA. If the client supports P-256 ECDSA, the P-256 ECDSA certificate
  is preferred over RSA. The certificate that it falls back to might result in a failed handshake. For instance, a client only supports
  RSA certificates and the certificate only support ECDSA.
* The final selected certificate must adhere to the OCSP policy. If no such certificate is found, the connection is refused.

.. note::
   With the support of SNI-based certificate selection, it allows configuring large number of certificates for multiple hostnames.
   But if there is no matching SNI or there is OCSP policy failure on the matching SNI, it will look for the cert from the whole cert
   list(O(n) complexity), which could be a problem for a potential DoS attack. You could disable the full scan on sni mismatch with

>    :ref:`full_scan_certs_on_sni_mismatch <envoy_v3_api_field_extensions.transport_sockets.tls.v3.DownstreamTlsContext.full_scan_certs_on_sni_mismatch>`.

Reviewer 2: I would clarify this is disabled by default.

certificates. These may be a mix of RSA and P-256 ECDSA certificates for multiple services. The following rules apply:

Certificate config/loading rules:
* DNS SANs or Subject Common Name is extracted as server name pattern to match SNI during handshake.

> * Only one certificate of a particular type (RSA or ECDSA) may be specified for each server name pattern.

Reviewer: You'll need to clarify the behavior around wildcards. Is it valid or invalid to have a cert for `test.example.com` and also one for `*.example.com`? If both are valid, please specify which cert is chosen when SNI is `test.example.com`.

Author: Got it. Done.
